### PR TITLE
Add authorizer and include user for isFeatureEnable function

### PR DESCRIPTION
**Pull request by rohithbalaji123**
Part of #612 

### Description
Feed authorizer and user entity into Decision usecase to support permission based feature toggle in near future.

TODO:
- [x] Modify tests to inject authorizer

### CODE REVIEW

File: backend/app/usecase/feature/dynamic_test.go [hunk 1]
```diff
@@ -72,9 +74,9 @@ func TestDynamicDecisionMaker_IsFeatureEnable(t *testing.T) {
 			}()
 
 			ins := instrumentation.NewInstrumentation(lg, tm, mt, ana, ctxCh)
-			factory := NewDynamicDecisionMakerFactory(featureRepo)
+			factory := NewDynamicDecisionMakerFactory(featureRepo, authorizer.Authorizer{})
```
[the remark this answers is not in the document]
Comment (rohithbalaji123) [reply backend/app/usecase/feature/dynamic_test.go:77]: Yes, it will. I thought of updating along with the PR implementing the logic.

File: backend/app/usecase/feature/dynamic.go [hunk 2]
```diff
@@ -32,13 +34,14 @@ var _ DecisionMakerFactory = (*DynamicDecisionMakerFactory)(nil)
 // DynamicDecisionMakerFactory creates feature decision maker.
 type DynamicDecisionMakerFactory struct {
 	featureToggleRepo repository.FeatureToggle
+	authorizer        authorizer.Authorizer
```
[the remark this answers is not in the document]
Comment (rohithbalaji123) [reply backend/app/usecase/feature/dynamic.go:37]: Okay, I'll add the implementation in the same PR.

**Comment by rohithbalaji123** [comment]
I have added implementation for decision making.
The PR is ready for review.

File: backend/app/adapter/routing/handle.go [hunk 3]
```diff
@@ -101,3 +106,28 @@ func FeatureHandle(
...
+func getUser(r *http.Request, authenticator authenticator.Authenticator) *entity.User {
+	authToken := getBearerToken(r)
+	user, err := authenticator.GetUser(authToken)
+	if err != nil {
+		return nil
+	}
+	return &user
+}
+
+// getBearerToken parses Authorization token of format "Bearer <token>"
```
Comment (byliuyang) [inline backend/app/adapter/routing/handle.go:119]:
```suggestion
// getBearerToken parses Authorization token with format "Bearer <token>"
```

File: backend/app/usecase/feature/decision.go [hunk 4]
```diff
@@ -1,11 +1,24 @@
 package feature
 
-import "github.com/short-d/short/backend/app/usecase/instrumentation"
+import (
+	"github.com/short-d/short/backend/app/entity"
+	"github.com/short-d/short/backend/app/usecase/instrumentation"
+)
+
+// PermissionChecker type is the signature for the checker method used to check
```
Comment (byliuyang) [inline backend/app/usecase/feature/decision.go:8]:
```suggestion
// PermissionChecke checks
```

File: backend/app/usecase/feature/decision.go [hunk 5]
```diff
@@ -1,11 +1,24 @@
...
+// if the user has the privilege to access a feature
```
Comment (byliuyang) [inline backend/app/usecase/feature/decision.go:9]:
```suggestion
// whether the user is allowed to access a feature
```

File: backend/app/usecase/feature/decision.go (hunk 4 above)
Comment (byliuyang) [reply backend/app/usecase/feature/decision.go:8]: Here is Golang comment best practices: https://blog.golang.org/godoc

File: backend/app/usecase/feature/decision.go [hunk 6]
```diff
@@ -1,11 +1,24 @@
...
+type PermissionChecker func(user entity.User) (bool, error)
+
+type PermissionToggle string
+
+const (
+	AdminPanel PermissionToggle = "admin-panel"
+)
```
Comment (byliuyang) [inline backend/app/usecase/feature/decision.go:16]: I think can be separated to `checker.go` file

File: backend/app/usecase/feature/dynamic_test.go [hunk 7]
```diff
@@ -46,11 +52,66 @@ func TestDynamicDecisionMaker_IsFeatureEnable(t *testing.T) {
...
 					ID:        "example-feature",
 					IsEnabled: true,
+					Type:      entity.ManualToggle,
 				},
 			},
 			featureID:         "example-feature",
 			expectedIsEnabled: true,
 		},
+		{
+			name: "permission not found",
```
Comment (byliuyang) [inline backend/app/usecase/feature/dynamic_test.go:62]:
```suggestion
			name: "no permission",
```

File: backend/app/usecase/feature/dynamic_test.go [hunk 8]
```diff
@@ -46,11 +52,66 @@ func TestDynamicDecisionMaker_IsFeatureEnable(t *testing.T) {
...
+			roles: map[string][]role.Role{
+				"id": {role.Basic},
+			},
+			user: entity.User{
+				ID: "id",
+			},
+			expectedIsEnabled: false,
+		},
+		{
+			name: "permission disabled",
```
Comment (byliuyang) [inline backend/app/usecase/feature/dynamic_test.go:80]: I am curious what's this? I saw in the lines below, `IsEnabled` is true.

Comment (byliuyang) [reply backend/app/usecase/feature/dynamic_test.go:80]: User should also be admin!

File: backend/app/usecase/feature/dynamic_test.go [hunk 9]
```diff
@@ -46,11 +52,66 @@ func TestDynamicDecisionMaker_IsFeatureEnable(t *testing.T) {
...
+			roles: map[string][]role.Role{
+				"id": {role.Basic},
+			},
+			user: entity.User{
+				ID: "id",
+			},
+			expectedIsEnabled: false,
+		},
+		{
+			name: "permission enabled",
```
Comment (byliuyang) [inline backend/app/usecase/feature/dynamic_test.go:98]:
```suggestion
			name: "has permission",
```

File: backend/app/usecase/feature/dynamic_test.go [hunk 10]
```diff
@@ -71,10 +132,14 @@ func TestDynamicDecisionMaker_IsFeatureEnable(t *testing.T) {
...
+			fakeRolesRepo := repository.NewUserRoleFake(testCase.roles)
+			rb := rbac.NewRBAC(fakeRolesRepo)
+			au := authorizer.NewAuthorizer(rb)
+
 			ins := instrumentation.NewInstrumentation(lg, tm, mt, ana, ctxCh)
-			factory := NewDynamicDecisionMakerFactory(featureRepo)
+			factory := NewDynamicDecisionMakerFactory(featureRepo, au)
 			decision := factory.NewDecision(ins)
-			gotIsEnabled := decision.IsFeatureEnable(testCase.featureID)
+			gotIsEnabled := decision.IsFeatureEnable(testCase.featureID, &testCase.user)
```
Comment (byliuyang) [inline backend/app/usecase/feature/dynamic_test.go:142]: I wonder whether we can add a test case when user is `nil`.

File: backend/app/usecase/feature/dynamic_test.go (hunk 8 above)
Comment (rohithbalaji123) [reply backend/app/usecase/feature/dynamic_test.go:80]: This is the condition for, "The feature is enabled, but the user is not allowed to access the feature."

Comment (rohithbalaji123) [reply backend/app/usecase/feature/dynamic_test.go:80]: I'll change the test name to `"feature enabled, does not have permission"`

File: backend/app/usecase/feature/dynamic_test.go [hunk 11]
```diff
@@ -46,11 +52,96 @@ func TestDynamicDecisionMaker_IsFeatureEnable(t *testing.T) {
...
+			toggles: map[string]entity.Toggle{
+				"example-feature": {
+					ID:        "example-feature",
+					IsEnabled: true,
+					Type:      entity.PermissionToggle,
+				},
+			},
+			featureID: "example-feature",
+			roles: map[string][]role.Role{
+				"id": {role.Basic},
```
Comment (byliuyang) [inline backend/app/usecase/feature/dynamic_test.go:72]:
```suggestion
				"id": {role.Admin},
```
We need to make sure that the feature is not shown only because of the toggle is not enabled.

File: backend/app/usecase/feature/dynamic_test.go [hunk 12]
```diff
@@ -46,11 +52,96 @@ func TestDynamicDecisionMaker_IsFeatureEnable(t *testing.T) {
...
+			},
+			user: entity.User{
+				ID: "id",
+			},
+			expectedIsEnabled: false,
+		},
+		{
+			name: "permission toggle, feature enabled, user has permission",
```
Comment (byliuyang) [inline backend/app/usecase/feature/dynamic_test.go:98]:
```suggestion
			name: "permission toggle enabled, user has permission",
```

File: backend/app/usecase/feature/dynamic_test.go [hunk 13]
```diff
@@ -46,11 +52,96 @@ func TestDynamicDecisionMaker_IsFeatureEnable(t *testing.T) {
...
+			roles: map[string][]role.Role{
+				"id": {role.Admin},
+			},
+			user: entity.User{
+				ID: "id",
+			},
+			expectedIsEnabled: true,
+		},
+		{
+			name: "permission toggle, feature disabled, user has permission",
```
Comment (byliuyang) [inline backend/app/usecase/feature/dynamic_test.go:116]:
```suggestion
			name: "permission toggle disabled, user has permission",
```

File: backend/app/usecase/feature/dynamic.go [hunk 14]
```diff
@@ -10,45 +12,93 @@ var _ DecisionMaker = (*DynamicDecisionMaker)(nil)
...
-	f.instrumentation.MadeFeatureDecision(featureID, toggle.IsEnabled)
+	d.instrumentation.FeatureToggleRetrievalSucceed()
+
+	decision := d.makeDecision(toggle, user)
+	d.instrumentation.MadeFeatureDecision(featureID, decision)
+	return decision
+}
+
+func (d DynamicDecisionMaker) makeDecision(toggle entity.Toggle, user *entity.User) bool {
+	switch toggle.Type:
```
Comment (byliuyang) [inline backend/app/usecase/feature/dynamic.go:36]:
```suggestion
	switch toggle.Type {
```

File: backend/app/usecase/feature/dynamic_test.go [hunk 15]
```diff
@@ -46,11 +52,96 @@ func TestDynamicDecisionMaker_IsFeatureEnable(t *testing.T) {
...
+					IsEnabled: true,
+					Type:      entity.PermissionToggle,
+				},
+			},
+			featureID: "admin-panel",
+			roles: map[string][]role.Role{
+				"id": {role.Admin},
+			},
+			user: entity.User{
+				ID: "id",
```
Comment (byliuyang) [inline backend/app/usecase/feature/dynamic_test.go:111]:
```suggestion
				ID: "alpha",
```

File: backend/app/usecase/feature/dynamic_test.go [hunk 16]
```diff
@@ -46,11 +52,96 @@ func TestDynamicDecisionMaker_IsFeatureEnable(t *testing.T) {
...
+					IsEnabled: false,
+					Type:      entity.PermissionToggle,
+				},
+			},
+			featureID: "admin-panel",
+			roles: map[string][]role.Role{
+				"id": {role.Admin},
+			},
+			user: entity.User{
+				ID: "id",
```
Comment (byliuyang) [inline backend/app/usecase/feature/dynamic_test.go:129]:
```suggestion
				ID: "alpha",
```
